extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/OffHeapNamespaceExtractionCacheManager.java

```diff
@@ -19,47 +19,113 @@
 
 package io.druid.server.lookup.namespace.cache;
 
-import com.google.common.base.Preconditions;
 import com.google.common.base.Throwables;
-import com.google.common.util.concurrent.Striped;
 import com.google.inject.Inject;
 import com.metamx.emitter.service.ServiceEmitter;
 import com.metamx.emitter.service.ServiceMetricEvent;
-
 import io.druid.java.util.common.lifecycle.Lifecycle;
 import io.druid.java.util.common.logger.Logger;
-import io.druid.query.lookup.namespace.ExtractionNamespace;
-import io.druid.query.lookup.namespace.ExtractionNamespaceCacheFactory;
 import org.mapdb.DB;
 import org.mapdb.DBMaker;
+import org.mapdb.HTreeMap;
+import sun.misc.Cleaner;
 
 import java.io.File;
 import java.io.IOException;
-import java.util.Map;
-import java.util.UUID;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.locks.Lock;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  *
  */
 public class OffHeapNamespaceExtractionCacheManager extends NamespaceExtractionCacheManager
 {
   private static final Logger log = new Logger(OffHeapNamespaceExtractionCacheManager.class);
+
+  private class MapDbCacheDisposer implements Runnable
+  {
+    final String mapDbKey;
+    /**
+     * Manages the race between dispose via {@link #disposeManually()} and automatic dispose by the JVM via
+     * {@link #run()}.
+     */
+    final AtomicBoolean disposed = new AtomicBoolean(false);
+
+    private MapDbCacheDisposer(String mapDbKey)
+    {
+      this.mapDbKey = mapDbKey;
+    }
+
+    /**
+     * To be called by the JVM via {@link Cleaner#clean()}. The only difference from {@link #disposeManually()} is
+     * exception treatment.
+     */
+    @Override
+    public void run()
+    {
+      if (disposed.compareAndSet(false, true)) {
```

Comment:
I *think* this is OK if `doDispose` is still in process. Its guarded against running it again, and unlike a `stop` kind of method, either call eventually needs to be called, and if the JVM is shutdown before it is called then it doesn't really matter.